Flat Extraction Feature

mega5800 opened the pull request:
:+1::tada: First off, thanks for taking the time to contribute! :tada::+1:

Please fill out the following checklist:

- [x] Added tests for any bug fixes that changed existing code
- [x] Added tests for any new behavior
- [x] All unit tests are passing (please make sure all tests are passing for your branch/PR at https://ci.appveyor.com/project/activescott/lessmsi/history )


If you need any help at all, feel free to submit the PR and @-mention activescott and I'll be happy to assist!

Hello @activescott

I took care of this [ticket](https://github.com/activescott/lessmsi/issues/174).

I added new CLI commands "xfo" and "xfr".
"xfo" allows the user to flat extract MSI files while overwriting files with the same name.
"xfr" allows the user to flat extract MSI files while renaming files with the same name using the suffix "_" with a count.

As discussed, I experienced a few issues with creating test files for this new behaviour.
So I am sending this PR as you suggested.

Please let me know if you have any questions.

Thank you.

activescott commented on src/LessMsi.Cli/eExtractionMode.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace LessMsi.Cli
+{
+    public enum eExtractionMode
```
Can you please rename this to ExtractionMode per https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#naming-enumerations ?

mega5800 replied:
Will amend the code accordingly.

activescott commented on src/LessMsi.Cli/Program.cs:
```diff
@@ -124,10 +186,17 @@ private static void PrintProgress(IAsyncResult result)
+                sFileName = Path.Combine(Directory.GetCurrentDirectory(), sFileName);
+            }
+
+            if (isExtractionModeFlat(extractionMode))
```
Are you sure this is what you want in the case of !Path.IsPathRooted (the if above)? Now that this method has become more complicated it is probably worth a unit test with the various combinations of extractionMode and with scenarios where sFileName is rooted and when it is not rooted.

mega5800 replied:
How about I leave the `EnsureFileRooted` method as it was originally and extract my added logic into a separate method?
This way the logic in `EnsureFileRooted` method stays the same and I can call the newly created method whenever I need without involving `EnsureFileRooted` method.

activescott commented on src/LessMsi.Cli/eExtractionMode.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public enum eExtractionMode
+    {
+        None = 0,
```
You don't need to assign these numbers here these will be the default values. Do other parts of the code actually use the numeric value? Ideally other parts of the code just use the enum itself and do not rely on the numeric value anyway.

mega5800 replied:
Will amend the code accordingly.

activescott commented on src/LessMsi.Cli/Program.cs:
```diff
@@ -95,25 +101,81 @@ public static int Main(string[] args)
+
+        private static bool isExtractionModeFlat(eExtractionMode extractionMode)
+        {
+            bool flatExtractionModeFlag = extractionMode == eExtractionMode.OverwriteFlatExtraction;
+            flatExtractionModeFlag |= extractionMode == eExtractionMode.RenameFlatExtraction;
+
+            return flatExtractionModeFlag;
```
return eExtractionMode.OverwriteFlatExtraction || eExtractionMode.RenameFlatExtraction;

mega5800 replied:
Will amend the code accordingly.

mega5800 replied:
`return eExtractionMode.OverwriteFlatExtraction || eExtractionMode.RenameFlatExtraction;` 
does not make sense as the enum values `OverwriteFlatExtraction` and `RenameFlatExtraction` are not Boolean and does not check the value of the passed parameter `extractionMode`.

Can I instead offer the next expression:
`return extractionMode == ExtractionMode.RenameFlatExtraction || extractionMode == ExtractionMode.OverwriteFlatExtraction`

This expression compares the parameter `extractionMode` to the relevant `OverwriteFlatExtraction` and `RenameFlatExtraction` values.

activescott commented on src/LessMsi.Cli/Program.cs:
```diff
@@ -56,14 +58,18 @@ public static int Main(string[] args)
 				 * See https://github.com/mono/mono/blob/master/mcs/tools/mdoc/Mono.Documentation/mdoc.cs#L54  for an example of using "commands" and "subcommands" with the NDesk.Options lib.
 				 */
 
+                ExtractCommand extractCommandObject = new ExtractCommand();
```
remove the `...Object` postfix on the variable name pls.

mega5800 replied:
Will amend the code accordingly.

activescott commented on src/LessMsi.Cli/Program.cs:
```diff
@@ -95,25 +101,81 @@ public static int Main(string[] args)
+            {
+                Dictionary<string, int> fileNameCountingDict = new Dictionary<string, int>();
+
+                string originalOutDirName = outDirName.Remove(outDirName.Length - c_TempFolderSuffix.Length) + "\\";
```
It seems unnecessarily complicated to compute this new suffix name in one method (in `EnsureFileRooted`) and then to make assumption about that logic in this method later. So instead, I recommend that you just compute the outdir name in this `DoExtraction` method above and remove that logic from `EnsureFileRooted`. Then before you overrwrite the `outDir` local variable here, you can preserve it's original value in `originalOutDirName` in the same method. I think that will simplify the logic here. Let me know if I'm missing something!

mega5800 replied:
I agree with your approach here and will change the code accordingly.

activescott commented on src/LessMsi.Cli/Program.cs:
```diff
@@ -95,25 +101,81 @@ public static int Main(string[] args)
+                    if (fileNameCountingDict.ContainsKey(fileName))
+                    {
+                        fileSuffix = $"_{fileNameCountingDict[fileName]}";
+                        fileNameCountingDict[fileName]++;
```
Double check this line works the way you expect. Is the int value returned by the dict's get method returned by reference (I forget how .net works in this case)?

mega5800 replied:
Before sending this pull request, I did some manual testing with NUnit msi file.
I didn’t encounter any problems with renaming identical files with their appearance count.

activescott commented on src/LessMsi.Cli/Program.cs:
```diff
@@ -95,25 +101,81 @@ public static int Main(string[] args)
+
+            if (isExtractionModeFlat(extractionMode))
+            {
+                Dictionary<string, int> fileNameCountingDict = new Dictionary<string, int>();
```
This is a bit simpler to read and I believe you can use var here since it's local and it will infer the type from the initialization.
```suggestion
                var fileNameCountingDict = new Dictionary<string, int>();

```

activescott commented on src/LessMsi.Cli/Program.cs:
```diff
@@ -43,10 +43,12 @@ enum ConsoleReturnCode
-		/// The main entry point for the application.
-		/// </summary>
-		[STAThread]
+        private const string c_TempFolderSuffix = "_temp";
```
Per https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions

```suggestion
        private const string TempFolderSuffix = "_temp";

```

activescott commented on src/LessMsi.Cli/eExtractionMode.cs:
```diff
@@ -0,0 +1,10 @@
+    public enum eExtractionMode
+    {
+        None = 0,
+        RegularExtraction = 1,
```
Please add a `///` comment to each of these to document them. Maybe also rename `RegularExtraction` to a more descriptive name like `PreserveDirectories` or something

activescott commented:
@mega5800 Thank you for your contribution! I left a few comments that should help you simplify the code and some minor naming convention notes. 

To get the test working I think you can just go add new tests to CommandLineExtractTests.cs like [this one](https://github.com/activescott/lessmsi/blob/09e1d963ff1bc0a9fc61b1be49c7121e87566cf8/src/Lessmsi.Tests/CommandLineExtractTests.cs#L41). Just choose an existing small test file like the NUnit-2.5.2.9222.msi used in several tests there and make sure that the test function itself has a unique name and the `GetTestName()` function will take care of generating a path where it expects a test file. Then just run that test file without any "expected file" in the `src/Lessmsi.Tests/TestFiles/ExpectedOutput/...` directory. When you run that new test note these two things:
1. It will generate a new "actual output" file 
2. The error message when the test fails should show you the path to the generated file. 

Then you can copy the generated actual file to the location of the expected file (just be sure to verify that the output in that file is correct). Once you run the test again it should pass.

Let me know if that doesn't help - if it doesn't just try to write the test and add it to the PR and I'll take a look. Sorry I'm not more help right now but I'm on an ARM mac and struggling to get windows to run.

mega5800 commented:
Thank you @activescott for your comments regarding my pull request.

I have addressed most of your comments and marked them as resolved. 
Once I finish addressing all of them, I will send an updated pull request for your review.
After you approve the pull request, I will add the tests. 
I prefer not to start working on the testing code until all comments are covered.

Thank you.

mega5800 commented:
Hello @activescott.

I covered all your comments.
Please let me know if you have any further comments.

If you are content with the code, I will start working on the tests code.
Thank you.

mega5800 commented:
Hello @activescott.

For some reason, I was unable to produce the actual output files following your steps. 
I have added the flat extraction tests to this pull request and as you can see AppVeyor build failed because it doesn’t have the output files. 
I would be grateful for your assistance with this issue.

Thank you.

activescott commented:
Thanks @mega5800 ! I'll get looking into this tonight and figure out how to get those files generated. I must have forgotten something as I haven't added new ones in a while myself!

activescott commented on src/LessMsi.Cli/ExtractCommand.cs:
```diff
@@ -1,36 +1,53 @@
+            while (++i < args.Count)
+                filesToExtract.Add(args[i]);
+
+            Program.DoExtraction(msiFile, extractDir.TrimEnd('\"'), filesToExtract, getExtractionMode(allArgs[0]));
```
is it possible for `allArgs.Length < 1` here?

mega5800 replied:
Can you please elaborate?

activescott commented on src/Lessmsi.Tests/CommandLineExtractTests.cs:
```diff
@@ -16,21 +16,63 @@ public void Extract1Arg()
+        public void FlatOverwriteExtract2Args()
+        {
+            var commandLine = "xfo TestFiles\\MsiInput\\NUnit-2.5.2.9222.msi FlatOverwriteExtract2Args\\";
+            TestExtraction(commandLine, GetTestName(), "FlatOverwriteExtract2Args", false);
```
Okay check out [this line](https://ci.appveyor.com/project/activescott/lessmsi/builds/50041787#L489) in the CI output:

```
    LessMsi.Tests.CommandLineExtractTests.FlatOverwriteExtract2Args [FAIL]
      System.IO.FileNotFoundException : Could not find file 'C:\projects\lessmsi\src\.deploy\TestFiles\ExpectedOutput\FlatOverwriteExtract2Args.expected.csv'.
```

To get this test working locally try this:
1. Put an empty text file at the path that you checkout the `https://github.com/activescott/lessmsi/tree/master/src/Lessmsi.Tests/TestFiles/ExpectedOutput` folder from this repo. So something like: `c:\src\lessmsi\TestFiles\ExpectedOutput\FlatOverwriteExtract2Args.expected.csv` assuming you check it out at `c:\src\lessmsi\`
2. Then run that test again locally and I think it will fail and you'll see a link to the "actual" version of that file (it's in some subdirectory of the path where you run the tests). 
3. Confirm that the actual version of the file has the correct contents and then copy it over the "expected" path above. Then the test should pass.
4. Once the test passes and you're confident the output in the "expected" file is correct, commit the expected file you created/updated above to the repo and it should start passing in CI.

You should be able to follow this pattern for the other tests too.

mega5800 replied:
Hello @activescott.

I was able to create the actual and expected csv files.
However, when I run the test, I get the next error message:

```
LessMsi.Tests.CommandLineExtractTests.FlatOverwriteExtract2Args [FAIL]
      System.Exception : FileEntryGraph entries are not the equal: '\Addin.cs'!='\Addin.cs' at index '0'.'\addinsDialog.html'!='\addinsDialog.html' at index '1'.'\addinsDialog.jpg'!='\addinsDialog.jpg' at index '2'.'\agent.conf'!='\agent.conf' at index '3'.'\assembliesTab.jpg'!='\assembliesTab.jpg' at index '4'.'\AssemblyInfo.cpp'!='\AssemblyInfo.cpp' at index '5'.'\AssemblyInfo.cs'!='\AssemblyInfo.cs' at index '6'.'\AssemblyInfo.jsl'!='\AssemblyInfo.jsl' at index '7'.'\AssemblyInfo.vb'!='\AssemblyInfo.vb' at index '8'.'\assertions.html'!='\assertions.html' at index '9'.'\AssertSyntaxTests.cs'!='\AssertSyntaxTests.cs' at index '10'.'\AssertSyntaxTests.vb'!='\AssertSyntaxTests.vb' at index '11'.'\attributes.html'!='\attributes.html' at index '12'.'\bulletOff.gif'!='\bulletOff.gif' at index '13'.'\bulletOn.gif'!='\bulletOn.gif' at index '14'.'\category.html'!='\category.html' at index '15'.'\codeFuncs.js'!='\codeFuncs.js' at index '16'.'\collectionAssert.html'!='\collectionAssert.html' at index '17'.'\collectionConstraints.html'!='\collectionConstraints.html' at index '18'.'\combinatorial.html'!='\combinatorial.html' at index '19'.'\comparisonAsserts.html'!='\comparisonAsserts.html' at index '20'.'\comparisonConstraints.html'!='\comparisonConstraints.html' at index '21'.'\compoundConstraints.html'!='\compoundConstraints.html' at index '22'.'\conditionAsserts.html'!='\conditionAsserts.html' at index '23'.'\conditionConstraints.html'!='\conditionConstraints.html' at index '24'.'\configEditor.html'!='\configEditor.html' at index '25'.'\configEditor.jpg'!='\configEditor.jpg' at index '26'.'\configFiles.html'!='\configFiles.html' at index '27'.'\console-mock.jpg'!='\console-mock.jpg' at index '28'.'\consoleCommandLine.html'!='\consoleCommandLine.html' at index '29'.'\constraintModel.html'!='\constraintModel.html' at index '30'.'\contextMenu.html'!='\contextMenu.html' at index '31'.'\CoreExtensibility.sln'!='\CoreExtensibility.sln' at index '32'.'\cpp-cli-failures.build'!='\cpp-cli-failures.build' at index '33'.'\cpp-cli-failures.vcproj'!='\cpp-cli-failures.vcproj' at index '34'.'\cpp-cli-syntax.build'!='\cpp-cli-syntax.build' at index '35'.'\cpp-cli-syntax.cpp'!='\cpp-cli-syntax.cpp' at index '36'.'\cpp-cli-syntax.vcproj'!='\cpp-cli-syntax.vcproj' at index '37'.'\cpp-cli.sln'!='\cpp-cli.sln' at index '38'.'\cpp-managed-failures.build'!='\cpp-managed-failures.build' at index '39'.'\cpp-managed-failures.vcproj'!='\cpp-managed-failures.vcproj' at index '40'.'\cppsample.cpp'!='\cppsample.cpp' at index '41'.'\cppsample.h'!='\cppsample.h' at index '42'.'\cs-failures.build'!='\cs-failures.build' at index '43'.'\cs-failures.csproj'!='\cs-failures.csproj' at index '44'.'\cs-money.build'!='\cs-money.build' at index '45'.'\cs-money.csproj'!='\cs-money.csproj' at index '46'.'\cs-syntax.build'!='\cs-syntax.build' at index '47'.'\cs-syntax.csproj'!='\cs-syntax.csproj' at index '48'.'\CSharp.sln'!='\CSharp.sln' at index '49'.'\CSharpTest.cs'!='\CSharpTest.cs' at index '50'.'\culture.html'!='\culture.html' at index '51'.'\customConstraints.html'!='\customConstraints.html' at index '52'.'\datapoint.html'!='\datapoint.html' at index '53'.'\datapointProviders.html'!='\datapointProviders.html' at index '54'.'\delayedConstraint.html'!='\delayedConstraint.html' at index '55'.'\description.html'!='\description.html' at index '56'.'\directoryAssert.html'!='\directoryAssert.html' at index '57'.'\equalConstraint.html'!='\equalConstraint.html' at index '58'.'\equalityAsserts.html'!='\equalityAsserts.html' at index '59'.'\eventListeners.html'!='\eventListeners.html' at index '60'.'\exception.html'!='\exception.html' at index '61'.'\exceptionAsserts.html'!='\exceptionAsserts.html' at index '62'.'\explicit.html'!='\explicit.html' at index '63'.'\extensibility.html'!='\extensibility.html' at index '64'.'\extensionTips.html'!='\extensionTips.html' at index '65'.'\failure.png'!='\failure.png' at index '66'.'\favicon.ico'!='\favicon.ico' at index '67'.'\fileAssert.html'!='\fileAssert.html' at index '68'.'\fit-license.txt'!='\fit-license.txt' at index '69'.'\fit.dll'!='\fit.dll' at index '70'.'\fixtureSetup.html'!='\fixtureSetup.html' at index '71'.'\fixtureTeardown.html'!='\fixtureTeardown.html' at index '72'.'\generalOptions.jpg'!='\generalOptions.jpg' at index '73'.'\generalTab.jpg'!='\generalTab.jpg' at index '74'.'\getStarted.html'!='\getStarted.html' at index '75'.'\gui-screenshot.jpg'!='\gui-screenshot.jpg' at index '76'.'\gui-verify.jpg'!='\gui-verify.jpg' at index '77'.'\guiCommandLine.html'!='\guiCommandLine.html' at index '78'.'\identityAsserts.html'!='\identityAsserts.html' at index '79'.'\ignore.html'!='\ignore.html' at index '80'.'\ignored.png'!='\ignored.png' at index '81'.'\IMoney.cs'!='\IMoney.cs' at index '82'.'\IMoney.vb'!='\IMoney.vb' at index '83'.'\inconclusive.png'!='\inconclusive.png' at index '84'.'\index.html'!='\index.html' at index '85'.'\installation.html'!='\installation.html' at index '86'.'\jsharp-failures.build'!='\jsharp-failures.build' at index '87'.'\jsharp-failures.vjsproj'!='\jsharp-failures.vjsproj' at index '88'.'\jsharp.sln'!='\jsharp.sln' at index '89'.'\JSharpTest.jsl'!='\JSharpTest.jsl' at index '90'.'\langFilter.gif'!='\langFilter.gif' at index '91'.'\license.html'!='\license.html' at index '92'.'\license.txt'!='\license.txt' at index '93'.'\listMapper.html'!='\listMapper.html' at index '94'.'\loadtest-assembly.dll'!='\loadtest-assembly.dll' at index '95'.'\log4net.dll'!='\log4net.dll' at index '96'.'\logo.gif'!='\logo.gif' at index '97'.'\Logo.ico'!='\Logo.ico' at index '98'.'\mainMenu.html'!='\mainMenu.html' at index '99'.'\managed-cpp.sln'!='\managed-cpp.sln' at index '100'.'\maxtime.html'!='\maxtime.html' at index '101'.'\miniGui.jpg'!='\miniGui.jpg' at index '102'.'\Minimal.build'!='\Minimal.build' at index '103'.'\Minimal.cs'!='\Minimal.cs' at index '104'.'\Minimal.csproj'!='\Minimal.csproj' at index '105'.'\mock-assembly.dll'!='\mock-assembly.dll' at index '106'.'\Money.cs'!='\Money.cs' at index '107'.'\Money.vb'!='\Money.vb' at index '108'.'\MoneyBag.cs'!='\MoneyBag.cs' at index '109'.'\MoneyBag.vb'!='\MoneyBag.vb' at index '110'.'\MoneyTest.cs'!='\MoneyTest.cs' at index '111'.'\MoneyTest.vb'!='\MoneyTest.vb' at index '112'.'\multiAssembly.html'!='\multiAssembly.html' at index '113'.'\nonamespace-assembly.dll'!='\nonamespace-assembly.dll' at index '114'.'\nunit-agent.exe'!='\nunit-agent.exe' at index '115'.'\nunit-agent.exe.config'!='\nunit-agent.exe.config' at index '116'.'\nunit-console-runner.dll'!='\nunit-console-runner.dll' at index '117'.'\nunit-console-x86.exe'!='\nunit-console-x86.exe' at index '118'.'\nunit-console-x86.exe.config'!='\nunit-console-x86.exe.config' at index '119'.'\nunit-console.exe'!='\nunit-console.exe' at index '120'.'\nunit-console.exe.config'!='\nunit-console.exe.config' at index '121'.'\nunit-console.html'!='\nunit-console.html' at index '122'.'\nunit-console.tests.dll'!='\nunit-console.tests.dll' at index '123'.'\nunit-gui-runner.dll'!='\nunit-gui-runner.dll' at index '124'.'\nunit-gui.html'!='\nunit-gui.html' at index '125'.'\nunit-gui.tests.dll'!='\nunit-gui.tests.dll' at index '126'.'\nunit-x86.exe'!='\nunit-x86.exe' at index '127'.'\nunit-x86.exe.config'!='\nunit-x86.exe.config' at index '128'.'\nunit.core.dll'!='\nunit.core.dll' at index '129'.'\nunit.core.interfaces.dll'!='\nunit.core.interfaces.dll' at index '130'.'\nunit.core.tests.dll'!='\nunit.core.tests.dll' at index '131'.'\nunit.css'!='\nunit.css' at index '132'.'\nunit.exe'!='\nunit.exe' at index '133'.'\nunit.exe.config'!='\nunit.exe.config' at index '134'.'\nunit.fixtures.dll'!='\nunit.fixtures.dll' at index '135'.'\nunit.fixtures.tests.dll'!='\nunit.fixtures.tests.dll' at index '136'.'\nunit.framework.dll'!='\nunit.framework.dll' at index '137'.'\nunit.framework.tests.dll'!='\nunit.framework.tests.dll' at index '138'.'\nunit.framework.xml'!='\nunit.framework.xml' at index '139'.'\nunit.mocks.dll'!='\nunit.mocks.dll' at index '140'.'\nunit.mocks.tests.dll'!='\nunit.mocks.tests.dll' at index '141'.'\nunit.uiexception.dll'!='\nunit.uiexception.dll' at index '142'.'\nunit.uiexception.tests.dll'!='\nunit.uiexception.tests.dll' at index '143'.'\nunit.uikit.dll'!='\nunit.uikit.dll' at index '144'.'\nunit.uikit.tests.dll'!='\nunit.uikit.tests.dll' at index '145'.'\nunit.util.dll'!='\nunit.util.dll' at index '146'.'\nunit.util.tests.dll'!='\nunit.util.tests.dll' at index '147'.'\nunitAddins.html'!='\nunitAddins.html' at index '148'.'\NUnitFitTests.html'!='\NUnitFitTests.html' at index '149'.'\NUnitTests.config'!='\NUnitTests.config' at index '150'.'\NUnitTests.nunit'!='\NUnitTests.nunit' at index '151'.'\optionsDialog.jpg'!='\optionsDialog.jpg' at index '152'.'\pairwise.html'!='\pairwise.html' at index '153'.'\parameterizedTests.html'!='\parameterizedTests.html' at index '154'.'\pathConstraints.html'!='\pathConstraints.html' at index '155'.'\platform.html'!='\platform.html' at index '156'.'\pnunit-agent.exe'!='\pnunit-agent.exe' at index '157'.'\pnunit-agent.exe.config'!='\pnunit-agent.exe.config' at index '158'.'\pnunit-launcher.exe'!='\pnunit-launcher.exe' at index '159'.'\pnunit-launcher.exe.config'!='\pnunit-launcher.exe.config' at index '160'.'\pnunit.framework.dll'!='\pnunit.framework.dll' at index '161'.'\pnunit.html'!='\pnunit.html' at index '162'.'\projectEditor.html'!='\projectEditor.html' at index '163'.'\property.html'!='\property.html' at index '164'.'\propertyConstraint.html'!='\propertyConstraint.html' at index '165'.'\QuickStart.doc'!='\QuickStart.doc' at index '166'.'\quickStart.html'!='\quickStart.html' at index '167'.'\QuickStart.Spanish.doc'!='\QuickStart.Spanish.doc' at index '168'.'\random.html'!='\random.html' at index '169'.'\range.html'!='\range.html' at index '170'.'\ReadMe.txt'!='\ReadMe.txt' at index '171'.'\releaseDetail.html'!='\releaseDetail.html' at index '172'.'\releaseNotes.html'!='\releaseNotes.html' at index '173'.'\repeat.html'!='\repeat.html' at index '174'.'\requiredAddin.html'!='\requiredAddin.html' at index '175'.'\requiresMTA.html'!='\requiresMTA.html' at index '176'.'\requiresSTA.html'!='\requiresSTA.html' at index '177'.'\requiresThread.html'!='\requiresThread.html' at index '178'.'\Results.xsd'!='\Results.xsd' at index '179'.'\runFile.exe'!='\runFile.exe' at index '180'.'\runFile.exe.config'!='\runFile.exe.config' at index '181'.'\sameasConstraint.html'!='\sameasConstraint.html' at index '182'.'\SampleFixtureExtension.build'!='\SampleFixtureExtension.build' at index '183'.'\SampleFixtureExtension.cs'!='\SampleFixtureExtension.cs' at index '184'.'\SampleFixtureExtension.csproj'!='\SampleFixtureExtension.csproj' at index '185'.'\SampleFixtureExtensionAttribute.cs'!='\SampleFixtureExtensionAttribute.cs' at index '186'.'\SampleFixtureExtensionBuilder.cs'!='\SampleFixtureExtensionBuilder.cs' at index '187'.'\SampleFixtureExtensionTests.cs'!='\SampleFixtureExtensionTests.cs' at index '188'.'\SampleFixtureExtensionTests.csproj'!='\SampleFixtureExtensionTests.csproj' at index '189'.'\samples.common'!='\samples.common' at index '190'.'\samples.html'!='\samples.html' at index '191'.'\SampleSuiteExtension.build'!='\SampleSuiteExtension.build' at index '192'.'\SampleSuiteExtension.cs'!='\SampleSuiteExtension.cs' at index '193'.'\SampleSuiteExtension.csproj'!='\SampleSuiteExtension.csproj' at index '194'.'\SampleSuiteExtensionAttribute.cs'!='\SampleSuiteExtensionAttribute.cs' at index '195'.'\SampleSuiteExtensionBuilder.cs'!='\SampleSuiteExtensionBuilder.cs' at index '196'.'\SampleSuiteExtensionTests.cs'!='\SampleSuiteExtensionTests.cs' at index '197'.'\SampleSuiteExtensionTests.csproj'!='\SampleSuiteExtensionTests.csproj' at index '198'.'\sequential.html'!='\sequential.html' at index '199'.'\setCulture.html'!='\setCulture.html' at index '200'.'\settingsDialog.html'!='\settingsDialog.html' at index '201'.'\setup.html'!='\setup.html' at index '202'.'\setupFixture.html'!='\setupFixture.html' at index '203'.'\SimpleVBTest.vb'!='\SimpleVBTest.vb' at index '204'.'\skipped.png'!='\skipped.png' at index '205'.'\stringAssert.html'!='\stringAssert.html' at index '206'.'\stringConstraints.html'!='\stringConstraints.html' at index '207'.'\success.png'!='\success.png' at index '208'.'\suite.html'!='\suite.html' at index '209'.'\suiteBuilders.html'!='\suiteBuilders.html' at index '210'.'\Summary.xslt'!='\Summary.xslt' at index '211'.'\teardown.html'!='\teardown.html' at index '212'.'\test-assembly.dll'!='\test-assembly.dll' at index '213'.'\test-utilities.dll'!='\test-utilities.dll' at index '214'.'\test.conf'!='\test.conf' at index '215'.'\test.html'!='\test.html' at index '216'.'\testCase.html'!='\testCase.html' at index '217'.'\testcaseBuilders.html'!='\testcaseBuilders.html' at index '218'.'\testcaseProviders.html'!='\testcaseProviders.html' at index '219'.'\testCaseSource.html'!='\testCaseSource.html' at index '220'.'\testDecorators.html'!='\testDecorators.html' at index '221'.'\testFixture.html'!='\testFixture.html' at index '222'.'\testLoadOptions.jpg'!='\testLoadOptions.jpg' at index '223'.'\testOutputOptions.jpg'!='\testOutputOptions.jpg' at index '224'.'\testProperties.html'!='\testProperties.html' at index '225'.'\testProperties.jpg'!='\testProperties.jpg' at index '226'.'\TestResult.xml'!='\TestResult.xml' at index '227'.'\theory.html'!='\theory.html' at index '228'.'\throwsConstraint.html'!='\throwsConstraint.html' at index '229'.'\timeout.html'!='\timeout.html' at index '230'.'\timing-tests.dll'!='\timing-tests.dll' at index '231'.'\typeAsserts.html'!='\typeAsserts.html' at index '232'.'\typeConstraints.html'!='\typeConstraints.html' at index '233'.'\upgrade.html'!='\upgrade.html' at index '234'.'\utilityAsserts.html'!='\utilityAsserts.html' at index '235'.'\values.html'!='\values.html' at index '236'.'\valueSource.html'!='\valueSource.html' at index '237'.'\vb-failures.build'!='\vb-failures.build' at index '238'.'\vb-failures.vbproj'!='\vb-failures.vbproj' at index '239'.'\vb-money.build'!='\vb-money.build' at index '240'.'\vb-money.vbproj'!='\vb-money.vbproj' at index '241'.'\vb-samples.sln'!='\vb-samples.sln' at index '242'.'\vb-syntax.build'!='\vb-syntax.build' at index '243'.'\vb-syntax.vbproj'!='\vb-syntax.vbproj' at index '244'.'\vsSupport.html'!='\vsSupport.html' at index '245'.
```

I don't know why I get this error even though the actual files and the expected files are identical.
Maybe it has something to do with paths?
Anyway, would love to hear your opinion about this issue.

Thank you.

mega5800 replied:
@activescott, do you want me to include the generated test files in the pull request?

activescott replied:
Please only include the "expected" ones so that the tests pass.

activescott reviewed (changes requested):
I left a comment with steps that I think will help you get that test working. Tag me in a follow-up comment if it doesn't work! And if it does, request another review from me or tag me and we'll get this merged once those tests are passing.

mega5800 commented:
Hello @activescott.

I updated the testing logic to adapt for the flat extraction feature, and now I am able to run all the tests successfully.
However, I am facing some difficulties with Codacy errors.
I amended the code according to these comments, however I currently face some medium comments which are not relevant to the code IMO.

First one is about implementing IEquatable<T> interface to a struct I created for grouping file entry comparison bool value and possible error message.
Second one is about throwing exceptions by user code.
The only thing I updated here is the source of the error message, the exception was there.
Would love to get your help here.

Thank you.

activescott commented:
Great! I'll take a look tomorrow or the next day!

mega5800 commented:
@activescott any updates from your side?
Thank you.

activescott commented:
Sorry no update yet. I didn't have time this weekend. I'll look at it as soon as I can. Hopefully this week

activescott commented:
@mega5800 Check out 7e7de8dd4fa3f9ed827f9d905c08e202b7cb0070 and eec6c43e744521eda3740f3fae60fb47598b1e13 and let me know if you have questions or concerns. A couple of notes from my memory of what I saw:

- Use overloads rather than duplicating code.
- There used to be an "equality pattern" or comparable pattern or something in C# that I recalled from how to implement comparisons for value types (structs). That has evolved since I was deep into C# but this looks like a more recent explanation of how that all works: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/how-to-define-value-equality-for-a-type

Please do give it a good review as it is late for me and I was hacking my way through a bit quickly :) I also didn't review your latest changes thoroughly. Why don't you look at this and make sure you feel good about it and if you do I'll take one last look before we merge.

activescott reviewed (changes requested):
@mega5800 I just realized we need to also update https://github.com/activescott/lessmsi/blob/master/src/LessMsi.Cli/ShowHelpCommand.cs to include the new options options.

Please update there, and I'll use that help and docs in wiki per #199 to write the proper commit message on merge and it will show this information in the release notes.

mega5800 commented:
@activescott Just updated the help text

activescott commented:
:tada: This PR is included in version 2.1.0 :tada:

The release is available on:
- [GitHub release](https://github.com/activescott/lessmsi/releases/tag/v2.1.0)
- `v2.1.0`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
